### PR TITLE
Support `__rmod__`

### LINKED ISSUE #58035: Support `__rmod__`
## 🚀 Feature
Support `aten::remainder` for Scalar x Tensor inputs (`aten::remainder(Scalar, Tensor)`), and `torch.Tensor.__rmod__`.
(cc: @mruberry, @rgommers, @emcastillo and @kmaehashi)

## Motivation and Pitch

To enhance the compatibility with NumPy’s interface (c.f. https://github.com/pytorch/pytorch/issues/38349).

```
>>> 5 % numpy.array([1, 2, 3, 4, 5])
array([0, 1, 2, 1, 0])
>>> 5 % torch.tensor([1, 2, 3, 4, 5])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for %: 'int' and 'Tensor'
```

## References

* [NumPy numpy.remainder]https://numpy.org/doc/stable/reference/generated/numpy.remainder.html)
* [Python Array API: remainder(x1, x2, /)](https://data-apis.org/array-api/latest/API_specification/elementwise_functions.html#remainder-x1-x2)
* [Python Array API: Reflected Operators](https://data-apis.org/array-api/latest/API_specification/array_object.html#reflected-operators)

### PR DESCRIPTION
**Pull request by asi1024**
Fixes #58035.

This PR implements `torch.Tensor.__rmod__` and `torch.remainder(scalar, tensor)` for the compatibility with NumPy’s interface.
(cc: @mruberry, @rgommers, @emcastillo, @kmaehashi)

TODO:
  - [x] Update `tensor_binary_op` in test/test_binary_ufuncs.py after #58216 is merged.

### CODE REVIEW

**Review by mruberry — COMMENTED** [review]
Cool! This is great Python Array API compatibility, @asi1024

This is going to need a "forward" test to validate that the operation computes what it should when give a scalar x tensor input.

We have some tests for remainder here: 

https://github.com/pytorch/pytorch/blob/470cd6451400d9c96d20b91e3915172abc056455/test/test_binary_ufuncs.py#L1638

That can probably be extended

cc @rgommers, looks like the Python Array API doesn't elaborate on rmod, but it might want to be more explicit about whether it should call fmod or remainder. NumPy's rmod looks like it calls remainder, so I assume that's what's intended?

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -5352,6 +5352,17 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
                         device_type='cuda', dtypes=(torch.bfloat16,)),)),
+    OpInfo('__rmod__',
+           op=torch.Tensor.__rmod__,
+           dtypes=all_types_and(torch.bfloat16, torch.half),
+           dtypesIfCPU=floating_types_and(torch.half,),
+           dtypesIfCUDA=all_types_and(torch.bfloat16, torch.half, torch.bool),
+           sample_inputs_func=sample_inputs_rbinops,
+           supports_out=False,
+           skips=(SkipInfo('TestCommon', 'test_variant_consistency_jit',),),
+           supports_autograd=False,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5363]: Interesting... why doesn't this support autograd? torch.remainder supports autograd typically, right?

**Comment by rgommers** [comment]
> cc @rgommers, looks like the Python Array API doesn't elaborate on rmod, but it might want to be more explicit about whether it should call fmod or remainder. NumPy's rmod looks like it calls remainder, so I assume that's what's intended?

`__mod__` has a note saying it uses `remainder`: https://data-apis.org/array-api/latest/API_specification/array_object.html#mod-self-other. So `__rmod__` should do the same.

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (asi1024) [reply torch/testing/_internal/common_methods_invocations.py:5363]: `torch.remainder` seems not to support autograd of the second argument currently.
```py
>>> x1 = torch.tensor([3.], requires_grad=True)
>>> x2 = torch.tensor([5.], requires_grad=True)
>>> y = x1 % x2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: the derivative for 'other' is not implemented
```

**Comment by asi1024** [comment]
> This is going to need a "forward" test to validate that the operation computes what it should when give a scalar x tensor input.

@mruberry I added torch.remainder(Scalar, Tensor) forward test now!

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:5362]: Interesting. Can we mark the operation as supporting autograd but, when `requires_grad=True`, mark the second input as not requiring gradient still? We can add a comment explaining why.

Or is there some reason why that wouldn't work?

File: test/test_binary_ufuncs.py [hunk 2]
```diff
@@ -1637,30 +1637,30 @@ def test_fmod_remainder_by_zero_integral(self, device, dtype):
     @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=False, include_bool=False, include_complex=False))
     def test_fmod_remainder(self, device, dtype):
         # Use numpy as reference
-        def _helper(x, mod):
-            fns_list = ((torch.fmod, torch.Tensor.fmod_, np.fmod),
-                        (torch.remainder, torch.Tensor.remainder_, np.remainder))
+        def _helper(x, mod, fns_list):
             for fn, inplace_fn, ref_fn in fns_list:
-                np_x = x.cpu().numpy()
+                np_x = x.cpu().numpy() if torch.is_tensor(x) else x
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1642]: Nice update

File: test/test_binary_ufuncs.py [hunk 3]
```diff
@@ -1677,7 +1677,14 @@ def _helper(x, mod):
             mods.append(mod_float)
 
         for dividend, mod in product([x, x.t()], mods):
-            _helper(dividend, mod)
+            _helper(dividend, mod,
+                    ((torch.fmod, torch.Tensor.fmod_, np.fmod),
+                     (torch.remainder, torch.Tensor.remainder_, np.remainder),))
+
+        for dividend, mod in product([5, 3.14], mods):
```
Comment (mruberry) [inline test/test_binary_ufuncs.py:1684]: Add a comment here that this tests the scalar x tensor case

**Review by mruberry — COMMENTED** [review]
This is looking good, @asi1024!; I have some small suggestions to add comments

Also, it looks like the tests didn't run, which is odd. Maybe pushing the comment updates will trigger them properly? Once we see the tests passing I think this is good to go!

**Comment by asi1024** [comment]
@mruberry CI seems to have failed in `TestNNCOpInfoCPU::test_unsupported___rmod___cpu_float32` test, but this test case is skipped in my local. Do you know what should I do to reproduce the test failure in my environment? Here is the test log of pytest.
```
test/test_jit_fuser_te.py::TestNNCOpInfoCPU::test_unsupported___rmod___cpu_float32 SKIPPED (Compiles with TensorExprKernel)
```

**Comment by mruberry** [comment]
> @mruberry CI seems to have failed in `TestNNCOpInfoCPU::test_unsupported___rmod___cpu_float32` test, but this test case is skipped in my local. Do you know what should I do to reproduce the test failure in my environment? Here is the test log of pytest.
> 
> ```
> test/test_jit_fuser_te.py::TestNNCOpInfoCPU::test_unsupported___rmod___cpu_float32 SKIPPED (Compiles with TensorExprKernel)
> ```

Before bothering to reproduce it, what about doing what it suggests and adding dunder rmod to the list here?

https://github.com/pytorch/pytorch/blob/60931611581f7d9fa8f40baee58533955d13b8ce/test/test_jit_fuser_te.py#L1811

**Comment by asi1024** [comment]
@mruberry Thanks! I fixed it as you commented and confirmed the test has passed!
The remaining CI failure is due to exceeding the time limit without output. Will this be green if CI is retriggered?

**Comment by mruberry** [comment]
> @mruberry Thanks! I fixed it as you commented and confirmed the test has passed!
> The remaining CI failure is due to exceeding the time limit without output. Will this be green if CI is retriggered?

Thank you, @asi1024. And don't worry about the remaining failure: it's unrelated. Also, if I don't respond to you in two business days please ping me. Sometimes notifications are lost. I'm sorry you waited so long for a response. I'll review the PR now, too.

**Review by mruberry — APPROVED** [review]
Great work, @asi1024! And PyTorch is one step closer to supporting the Python Array API!

**Comment by asi1024** [comment]
@mruberry Thanks! :smiley:

**Comment by mruberry** [comment]
Interestingly this hit an internal error:

```
stderr: caffe2/aten/src/ATen/native/BinaryOps.cpp:633:10: error: no matching function for call to 'remainder'
  return native::remainder(wrapped_scalar_tensor(self), other);
         ^~~~~~~~~~~~~~~~~
stderr: caffe2/aten/src/ATen/native/BinaryOps.cpp:633:10: error: no matching function for call to 'remainder'
  return native::remainder(wrapped_scalar_tensor(self), other);
```

maybe change the namespace to `at`?

**Comment by asi1024** [comment]
Fixed the namespace and rebase, but the CI failed with rocm build failure. The message `No problems were identified.` is left. Could you help me figure out why it's failing?

**Comment by mruberry** [comment]
No worries about the ROCm failure, it's in the base. We fixed it yesterday. Running this through internal tests now.

**Comment by mruberry** [comment]
Internal tests look good; landing this now

**Comment by asi1024** [comment]
@mruberry Thanks for your review!

**Pull request closed without merging**